lib/rucio/core/rule.py

```diff
@@ -2394,6 +2395,87 @@ def release_parent_rule(child_rule_id, remove_parent_expiration=False, *, sessio
+def list_rules_for_rse_decommissioning(
+    rse_id: str,
+    *,
+    session: "Session"
+) -> "Iterable[dict[str, Any]]":
+    """Return a generator of rules at the RSE that is being decommissioned.
+
+    Decommissioning of an RSE involves deleting or moving away all rules that are
+    locking the replicas that exist at the RSE. The rules can be enforcing
+    dataset-level and/or file-level locks. Because rules are defined in terms of
+    RSE expressions, we need to first identify the locks with the RSE id and make
+    the list of rules that are enforcing such locks.
+    This function starts by listing the rules for dataset-level locks, and
+    proceeds to file-level locks once all of the former have been removed. This
+    two-step implementation is motivated by the fact that there can be a large
+    number of file-level locks resulting from the same dataset-level rule.
+
+    :param rse_id: Id of the RSE being decommissioned.
+    :param session: The database session in use.
+    :returns: A generator that yields rule dictionaries.
+    """
+    # Get rules with dataset locks first.
+    list_stmt = select(
+        models.ReplicationRule
+    ).join(
+        models.DatasetLock,
+        models.ReplicationRule.id == models.DatasetLock.rule_id
+    ).where(
+        models.DatasetLock.rse_id == rse_id
+    )
+
+    # Test: one-go execution
+#     has_dataset_rules = False
+
+    for rule in session.execute(list_stmt).yield_per(5).scalars():
+        # Test: one-go execution
+#         has_dataset_rules = True
+        yield rule.to_dict()
+
+    # Test: one-go execution
+#     if has_dataset_rules:
+#         # If rules on datasets exist, there can be thousands of replica locks
+#         # that refer to identical (dataset-level) rules. We therefore exit
+#         # here if there is any dataset rule.
+#         return
+
+    # ReplicaLock ("locks") table is not indexed by RSE ID, so we instead go
+    # through the RSEFileAssociation ("replicas") table.
+    # FIXME This query is too slow - consulting the DBAs (Y.I. & D.C. Dec 21 2022)
+
+    # Test: one-go execution
+    dataset_rule_ids = list_stmt.subquery()
+
+    list_stmt = select(
+        models.ReplicationRule
+    ).with_hint(
+        models.RSEFileAssociation,
+        "INDEX(REPLICAS  REPLICAS_RSE_ID_IDX)",
+        "oracle"
+    ).with_hint(
+        models.ReplicaLock,
+        "INDEX(LOCKS LOCKS_RULE_ID_IDX)",
+        "oracle"
+    ).join(
+        models.ReplicaLock,
+        models.ReplicationRule.id == models.ReplicaLock.rule_id
+    ).join(
+        models.RSEFileAssociation,
+        and_(models.ReplicaLock.scope == models.RSEFileAssociation.scope,
+             models.ReplicaLock.name == models.RSEFileAssociation.name,
+             models.ReplicaLock.rse_id == models.RSEFileAssociation.rse_id)
+    ).where(
+        models.RSEFileAssociation.rse_id == rse_id,
+        models.ReplicaLock.rule_id.not_in(dataset_rule_ids) # Test: one-go execution
+    )
```

Comment:
This took quite a bit of effort (and assistance from experts) to get right, but I think we have it:

```python
query_rules_from_replicas = select(
    models.ReplicationRule
).prefix_with(
    '/*+ USE_NL(locks) LEADING(replicas locks) */',
    'oracle'
).distinct(
).join_from(
    models.RSEFileAssociation,
    models.ReplicaLock,
    and_(models.RSEFileAssociation.scope == models.ReplicaLock.scope,
         models.RSEFileAssociation.name == models.ReplicaLock.name,
         models.RSEFileAssociation.rse_id == models.ReplicaLock.rse_id)
).join(
    models.ReplicationRule,
    models.ReplicaLock.rule_id == models.ReplicationRule.id
).where(
    models.RSEFileAssociation.rse_id == rse_id,
    models.ReplicaLock.rule_id.not_in(
        query_rules_from_dataset_locks.with_only_columns(models.ReplicationRule.id)
    )
)
```

Please replace and adjust the surrounding code and documentation as necessary. I think we have to introduce a third query between the two, but can do in the future.

Reply:
Done. Please let me know about the third query when necessary.

Reply:
The third query can be done post-merge. For now, to complete this item, please adjust the documentation string (the last few sentences feel outdated).

Reply:
Without introducing trailing whitespace, please. :smile:

Reply:
Argh, sorry! I usually use an editor that highlights trailing whitespaces, and forgot that the one I use on lxplus doesn't..